70_FLIPFluids/app_resources/cellUtils.hlsl

```diff
@@ -0,0 +1,182 @@
+{
+    cellMaterials = (cellMaterials & ~CellMatMask) | ((cellMaterial << CellMatMaskShift) & CellMatMask);
+}
+inline uint3 getCellMaterial(uint3 cellMaterials)
+{
+    return (cellMaterials & CellMatMask) >> CellMatMaskShift;
+}
+inline void setXPrevMaterial(inout uint3 cellMaterials, uint3 cellMaterial)
+{
+    cellMaterials = (cellMaterials & ~XPrevMatMask) | ((cellMaterial << XPrevMatMaskShift) & XPrevMatMask);
+}
+inline uint3 getXPrevMaterial(uint3 cellMaterials)
+{
+    return (cellMaterials & XPrevMatMask) >> XPrevMatMaskShift;
+}
+inline void setXNextMaterial(inout uint3 cellMaterials, uint3 cellMaterial)
+{
+    cellMaterials = (cellMaterials & ~XNextMatMask) | ((cellMaterial << XNextMatMaskShift) & XNextMatMask);
+}
+inline uint3 getXNextMaterial(uint3 cellMaterials)
+{
+    return (cellMaterials & XNextMatMask) >> XNextMatMaskShift;
+}
+inline void setYPrevMaterial(inout uint3 cellMaterials, uint3 cellMaterial)
+{
+    cellMaterials = (cellMaterials & ~YPrevMatMask) | ((cellMaterial << YPrevMatMaskShift) & YPrevMatMask);
+}
+inline uint3 getYPrevMaterial(uint3 cellMaterials)
+{
+    return (cellMaterials & YPrevMatMask) >> YPrevMatMaskShift;
+}
+inline void setYNextMaterial(inout uint3 cellMaterials, uint3 cellMaterial)
+{
+    cellMaterials = (cellMaterials & ~YNextMatMask) | ((cellMaterial << YNextMatMaskShift) & YNextMatMask);
+}
+inline uint3 getYNextMaterial(uint3 cellMaterials)
+{
+    return (cellMaterials & YNextMatMask) >> YNextMatMaskShift;
+}
+inline void setZPrevMaterial(inout uint3 cellMaterials, uint3 cellMaterial)
+{
+    cellMaterials = (cellMaterials & ~ZPrevMatMask) | ((cellMaterial << ZPrevMatMaskShift) & ZPrevMatMask);
+}
+inline uint3 getZPrevMaterial(uint3 cellMaterials)
+{
+    return (cellMaterials & ZPrevMatMask) >> ZPrevMatMaskShift;
+}
+inline void setZNextMaterial(inout uint3 cellMaterials, uint3 cellMaterial)
+{
+    cellMaterials = (cellMaterials & ~ZNextMatMask) | ((cellMaterial << ZNextMatMaskShift) & ZNextMatMask);
+}
+inline uint3 getZNextMaterial(uint3 cellMaterials)
+{
+    return (cellMaterials & ZNextMatMask) >> ZNextMatMaskShift;
+}
+
+
+inline bool isSolidCell(uint cellMaterial)
+{
+    return cellMaterial == CM_SOLID;
+}
+inline bool isFluidCell(uint cellMaterial)
+{
+    return cellMaterial == CM_FLUID;
+}
+inline bool isAirCell(uint cellMaterial)
+{
+    return cellMaterial == CM_AIR;
+}
+
+inline bool3 isSolidCell(uint3 cellMaterial)
+{
+    return cellMaterial == (uint3)CM_SOLID;
+}
+inline bool3 isFluidCell(uint3 cellMaterial)
+{
+    return cellMaterial == (uint3)CM_FLUID;
+}
+inline bool3 isAirCell(uint3 cellMaterial)
+{
+    return cellMaterial == (uint3)CM_AIR;
+}
+
+void enforceBoundaryCondition(inout float3 velocity, uint cellMaterial)
+{
+    bool3 is_solid_cell =
+        or((bool3)isSolidCell(getCellMaterial(cellMaterial)),
+        bool3(isSolidCell(getXPrevMaterial(cellMaterial)), isSolidCell(getYPrevMaterial(cellMaterial)), isSolidCell(getZPrevMaterial(cellMaterial))));
```

Comment:
general algorithm question, is this a staggered grid?
```
c --- s --- c
|               |
s              s
|               |
c --- s --- c
```

and the solid flags are kept between cells?

Reply:
more docs please, I don't get the axis-cell-material stuff, there seems to be a TON of redundant data